### ISSUE TITLE
bug: CORS Error when using APIGateway on localhost

### ISSUE THREAD

**Issue by AlexCoursi** [localstack/localstack#6877]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When I deploy a template SAM in localstack, the CORS are not disabled or took in account in the APIGateway service. 

I tried to put in my template: 
```
Globals:
    Api:
        Cors:
            AllowMethods: "'*'"
            AllowHeaders: "'*'"
            AllowOrigin: "'*'"
```

or:

```
ApiDeployment:
        Type: AWS::Serverless::Api
        Properties:
            Cors:
                AllowMethods: "'*'"
                AllowHeaders: "'*'"
                AllowOrigin: "'*'
```

I launched localstack with the following command:
`DATA_DIR=/tmp/localstack/data DEFAULT_REGION=eu-west-3 DEBUG=1 DISABLE_CORS_CHECKS=1 localstack start
`
Then, I create a simple page in react and I tried to call my APIGateway (using this URL: `http://localhost:4566/restapis/qj51i59y9u/dev/_user_request_`)

And I get an error about CORS with Method OPTIONS and POST:

`Blocage d’une requête multiorigines (Cross-Origin Request) : la politique « Same Origin » ne permet pas de consulter la ressource distante située sur http://localhost:4566/restapis/qj51i59y9u/dev/_user_request_/readdb. Raison : l’en-tête CORS « Access-Control-Allow-Origin » est manquant.`

It's working when I tried to call the local API with curl and it's working as well when I deploy on AWS, my simple page on local can call the API on AWS.

Before (around december 2021), it was working perfectly with localstack.

If I execute the following command it's working : 
`aws apigateway put-integration-response --rest-api-id oyhk02q4vs --resource-id r9k9d9cw1u --http-method OPTIONS --status-code 200 --selection-pattern 200 --response-parameters '{"method.response.header.Access-Control-Allow-Origin": "'"'"'*'"'"'","method.response.header.Access-Control-Allow-Methods": "'"'"'*'"'"'","method.response.header.Access-Control-Allow-Headers": "'"'"'*'"'"'"}' --endpoint-url=http://localhost:4566`

### Expected Behavior

I don't want to get CORS errors on localhost when I use:

`DATA_DIR=/tmp/localstack/data DEFAULT_REGION=eu-west-3 DEBUG=1 DISABLE_CORS_CHECKS=1 localstack start`

or

```
ApiDeployment:
        Type: AWS::Serverless::Api
        Properties:
            Cors:
                AllowMethods: "'*'"
                AllowHeaders: "'*'"
                AllowOrigin: "'*'
```

### How are you starting LocalStack?

With the `localstack` script

### Steps To Reproduce

- Create a template SAM with an APIGateway (POST Method)

- Deploy this template:

`samlocal deploy --guided --config-env dev`

- Create a simple page react ("create-react-app) and try to call the APIGateway

### Environment

```markdown
- OS:Ubuntu 20.04.5 LTS
- LocalStack:latest
```

**Comment by AlexCoursi** [comment]
Hello

I use the following template:

```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Parameters:
    Stage:
        Type: String
        Default: dev
    NomDuRoleLambda:
        Type: String
        Default: arn:aws:iam::XXXXXXXXXXXXXXX:role/Lambda_Role

Resources:
    myDB:
        Type: AWS::Serverless::SimpleTable
        Properties:
            TableName: !Sub "project-${Stage}"
            PrimaryKey:
                Name: id
                Type: String
            ProvisionedThroughput:
                ReadCapacityUnits: 5
                WriteCapacityUnits: 5

    LambdaRead:
        Type: AWS::Serverless::Function
        Properties:
            FunctionName: !Sub "project-read-${Stage}"
            CodeUri: functionRead/
            Handler: index.handler
            Runtime: nodejs14.x
            Role: !Sub "${NomDuRoleLambda}"
            Environment:
                Variables:
                    tableDynamo: !Ref myDB
            Events:
                apiForLambda:
                    Type: Api 
                    Properties:
                        Path: /readdb
                        Method: post
                        RestApiId: !Ref ApiDeployment
            Policies:
                DynamoDBReadPolicy:
                    TableName: !Ref myDB

    ApiDeployment:
        Type: AWS::Serverless::Api
        Properties:
            Auth:
                UsagePlan:
                    UsagePlanName: !Sub "project-${Stage}"
                    CreateUsagePlan: PER_API
                    Description: Pour appels post
            StageName: !Ref Stage
            Name: !Sub "project-api-${Stage}"
            Cors:
                AllowMethods: "'*'"
                AllowHeaders: "'*'"
                AllowOrigin: "'*'"


Outputs:
    ApiDeploymentOutput:
        Description: "API Gateway endpoint URL"
        Value: !Sub "https://${ApiDeployment}.execute-api.${AWS::Region}.amazonaws.com/${Stage}/"

```

The lambda code is the following:

```

var AWS = require('aws-sdk');

let options = {
    apiVersion: '2012-08-10'
}
if(typeof process.env.LOCALSTACK_HOSTNAME !== 'undefined' && process.env.LOCALSTACK_HOSTNAME !== null){
    options.endpoint = new AWS.Endpoint('http://'+process.env.LOCALSTACK_HOSTNAME+':4566')
}

var ddb = new AWS.DynamoDB(options);

exports.handler = async (event) => {
    try {

            var obj = JSON.parse(event.body);

            var ID = obj.id;
        
        
            var params = {
                TableName:process.env.tableDynamo,
                Key: {
                    id : {S: ID}
                }
            
            };
        
            var data;
            
            try{
                data = await ddb.getItem(params).promise();
                console.log("Item read successfully:", data);
            } catch(err){
                console.log("Error: ", err);
                data = err;
            }
        
        
            var response = {
                'statusCode': 200,
                'headers': {
                    "Access-Control-Allow-Headers" : "*",
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Methods": "*"
                },
                'body': JSON.stringify({
                    message: data
            })
        };
    } catch (err) {
        console.log(err);
        return err;
    }

    return response;
};

```

I'm not using Terraform, but maybe, TF is adding the parameters:

`aws apigateway put-integration-response --rest-api-id oyhk02q4vs --resource-id r9k9d9cw1u --http-method OPTIONS --status-code 200 --selection-pattern 200 --response-parameters '{"method.response.header.Access-Control-Allow-Origin": "'"'"'*'"'"'","method.response.header.Access-Control-Allow-Methods": "'"'"'*'"'"'","method.response.header.Access-Control-Allow-Headers": "'"'"'*'"'"'"}' --endpoint-url=http://localhost:4566`


Do you need more information ?

I know, in this case, a GET function would be better, but it's just a use case as an example for the dev team.

Regards
Alex

**Comment by AlexCoursi** [comment]
In your page, you add the following lines:

```terraform
resource "aws_api_gateway_integration_response" "options_integration_response" {
    rest_api_id   = aws_api_gateway_rest_api.cors_api.id
    resource_id   = aws_api_gateway_resource.cors_resource.id
    http_method   = aws_api_gateway_method.options_method.http_method
    status_code   = aws_api_gateway_method_response.options_200.status_code
    response_parameters = {
        "method.response.header.Access-Control-Allow-Headers" = "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'",
        "method.response.header.Access-Control-Allow-Methods" = "'GET,OPTIONS,POST,PUT'",
        "method.response.header.Access-Control-Allow-Origin" = "'*'"
    }
    depends_on = [aws_api_gateway_method_response.options_200]
```

With SAM, I don't need to do that and it's working on AWS. But with Localstack, it's not added automatically and it's not working. By adding theses lines, it's OK.

**Comment by MarcelStranak** [comment]
Hi @AlexCoursi can you please download the latest version of LocalStack (`docker pull localstack/localstack:latest`) and set the env variables like mentioned here https://docs.localstack.cloud/localstack/configuration/#security?

```sh
DISABLE_CORS_CHECKS=1
DISABLE_CUSTOM_CORS_APIGATEWAY=1
```

Please let me know if this helps.

**Comment by AlexCoursi** [comment]
Hello
I'm still getting the same error.
After updating the docker's images, localstack didn't restarted (problem with dynamo). 
With the following command, I succeded start localstack:

`docker run -it -p 4566:4566 -p 4571:4571 -e "DEBUG=1 DISABLE_CORS_CHECKS=1 DISABLE_CUSTOM_CORS_APIGATEWAY=1" localstack/localstack`

But the problem is still here.

My APIGateway (generated by localstack) is:

```
alex@alex-VirtualBox:~/Test/testAvecNode$ aws --endpoint-url=http://localhost:4566 apigateway get-resources --rest-api-id jfsl0bpfrq
{
    "items": [
        {
            "id": "7cyfdmva4d",
            "path": "/"
        },
        {
            "id": "vft0vnj33n",
            "parentId": "7cyfdmva4d",
            "pathPart": "readdb",
            "path": "/readdb",
            "resourceMethods": {
                "POST": {
                    "httpMethod": "POST",
                    "apiKeyRequired": false,
                    "methodResponses": {},
                    "methodIntegration": {
                        "type": "aws_proxy",
                        "httpMethod": "POST",
                        "uri": "arn:aws:apigateway:eu-west-3:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-3:000000000000:function:project-read-dev/invocations",
                        "requestParameters": {},
                        "requestTemplates": {},
                        "cacheKeyParameters": [],
                        "integrationResponses": {
                            "200": {
                                "statusCode": 200,
                                "responseTemplates": {}
                            }
                        }
                    }
                },
                "OPTIONS": {
                    "httpMethod": "OPTIONS",
                    "apiKeyRequired": false,
                    "methodResponses": {
                        "200": {
                            "statusCode": "200",
                            "responseParameters": {
                                "method.response.header.Access-Control-Allow-Origin": "'*'",
                                "method.response.header.Access-Control-Allow-Methods": "'*'",
                                "method.response.header.Access-Control-Allow-Headers": "'*'"
                            }
                        }
                    },
                    "methodIntegration": {
                        "type": "mock",
                        "httpMethod": "OPTIONS",
                        "requestParameters": {},
                        "requestTemplates": {
                            "application/json": "{\n  \"statusCode\" : 200\n}\n"
                        },
                        "cacheKeyParameters": [],
                        "integrationResponses": {
                            "200": {
                                "statusCode": "200",
                                "responseTemplates": {
                                    "application/json": "{}\n"
                                }
                            }
                        }
                    }
                }
            }
        }
    ]
}

```
Regards

**Comment by MarcelStranak** [comment]
Hi @AlexCoursi, can you please change a bit the docker command? I was testing it with active PRO license. 

`docker run -it -p 4566:4566 -p 4571:4571 -e DEBUG=1 -e DISABLE_CORS_CHECKS=1 -e DISABLE_CUSTOM_CORS_APIGATEWAY=1 localstack/localstack`

**Comment by AlexCoursi** [comment]
Hello !! 
Thank you so much, it's working !!
(Sorry for the late answer, Google sends Github's email as spams)
Regards
Alex

**Comment by AlexCoursi** [comment]
The solution:

`docker run -it -p 4566:4566 -p 4571:4571 -e DEBUG=1 -e DISABLE_CORS_CHECKS=1 -e DISABLE_CUSTOM_CORS_APIGATEWAY=1 localstack/localstack`

**Comment by wascou** [comment]
Ok, for others who might struggle with that one.

```
EXTRA_CORS_ALLOWED_ORIGINS="http://localhost:3000"
DISABLE_CORS_CHECKS="1"
DISABLE_CORS_HEADERS="0"
DISABLE_CUSTOM_CORS_APIGATEWAY="1"
```

1. I added in the authorized origins "http://localhost:3000", not only localhost

2. CORS check removes the extra checks from Localstack.
It implies to have both API GW with default pre flight and Lambda with response headers.

3. I disabled the CORS headers first but it was removing the headers sent back from API GW.
So the good option is to have it enabled.

4. And finally, Custom CORS are disabled, it let goes API GW to send their response headers.

Postman was helpful to check OPTIONS and request/response headers.
